fontir/src/orchestration.rs

```diff
@@ -34,83 +32,306 @@ bitflags! {
     }
 }
 
-pub type ContextItem<T> = Arc<RwLock<Option<Arc<T>>>>;
+impl Default for Flags {
+    /// Match the way gftools configures fontmake by default
+    fn default() -> Self {
+        Flags::EMIT_IR | Flags::PREFER_SIMPLE_GLYPHS
+    }
+}
 
-/// Generates fn $getter_name(&self) -> Arc<$value_type>.
+pub type ContextVec<T> = Arc<RwLock<Vec<T>>>;
+
+/// Clones are cheap and reference the same wrapped item courtesy of Arc
 ///
-/// Assumes we are in an impl block for a Context and that
-/// self.$lock_name is a ContextItem<$ir_type>
+/// Courtesy of Arc this is Clone even if T isn't
+#[derive(Default)]
+pub struct ContextItem<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    Ir: IrIo<I>,
+{
+    id: I,
+    acl: Arc<AccessControlList<I>>,
+    irio: Arc<Ir>,
+    value: Arc<RwLock<Option<Arc<T>>>>,
+}
+
+impl<I, T, Ir> ContextItem<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    Ir: IrIo<I>,
+    T: Persistable,
+{
+    pub fn new(id: I, acl: Arc<AccessControlList<I>>, irio: Arc<Ir>) -> Self {
+        ContextItem {
+            id,
+            acl,
+            irio,
+            value: Default::default(),
+        }
+    }
+
+    pub fn clone_with_acl(&self, acl: Arc<AccessControlList<I>>) -> Self {
+        ContextItem {
+            id: self.id.clone(),
+            acl,
+            irio: self.irio.clone(),
+            value: self.value.clone(),
+        }
+    }
+
+    /// Read item that you are sure must exist. Panic if not.
+    ///
+    /// Intended for use in [Work] to access items present in
+    /// [Work::read_access]. If these are missing something is horribly
+    /// wrong and we should kerplode.
+    pub fn get(&self) -> Arc<T> {
+        self.acl.assert_read_access(&self.id);
+        if let Some(in_memory) = self.try_get() {
+            return in_memory;
+        }
+
+        // it's *not* in memory but perhaps it's written down?
+        if self.irio.emit_ir() {
+            if let Some(mut reader) = self.irio.reader(&self.id) {
+                let restored = T::read(&mut reader);
+                *self.value.write() = Some(Arc::from(restored));
+            }            
+        }
+
+        // if we still don't have an answer just give up
+        self.try_get().unwrap_or_else(|| panic!("{:?} is not available", self.id))
+    }
+
+    /// Read an item that might not exist
+    pub fn try_get(&self) -> Option<Arc<T>> {
+        self.acl.assert_read_access(&self.id);
+        self.value.read().as_ref().cloned()
+    }
+
+    /// Update the value of the item whether or not it has changed.
+    ///
+    /// The change will be logged and anything relying on change detection, such as
+    /// conditional execution of dependent tasks, will fire.
+    ///
+    /// [ContextItem::set] is preferable where possible.
+    ///
+    /// This exists largely because write types in fontations do not always implement PartialEq.
+    /// TODO: should they?
```

Comment:
maybe? I have a mechanism for selectively implementing eq/hash/ord, but maybe it makes sense to just always implement these traits?

Reply:
I landed on maybe too! Deferred to https://github.com/googlefonts/fontations/issues/532